// @1inch/fusion-sdk/dist/cjs/api/quoter/quoter.request.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
var _limitordersdk = require("@1inch/limit-order-sdk");
var _quoterrequest = require("./quoter.request.js");
describe(__filename, function() {
    it('should return error if native currency', function() {
        expect(function() {
            return _quoterrequest.QuoterRequest.new({
                fromTokenAddress: _limitordersdk.Address.NATIVE_CURRENCY.toString(),
                toTokenAddress: '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2',
                amount: '1000000000000000000000',
                walletAddress: '0x00000000219ab540356cbb839cbe05303d7705fa',
                integratorFee: {
                    share: _limitordersdk.Bps.fromPercent(50),
                    receiver: _limitordersdk.Address.fromBigInt(10n),
                    value: new _limitordersdk.Bps(1n)
                }
            });
        }).toThrow(/wrap native currency/);
    });
    it('returns error fromTokenAddress or toTokenAddress equals ZERO_ADDRESS', function() {
        expect(function() {
            return _quoterrequest.QuoterRequest.new({
                fromTokenAddress: _limitordersdk.Address.ZERO_ADDRESS.toString(),
                toTokenAddress: '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2',
                amount: '1000000000000000000000',
                walletAddress: '0x00000000219ab540356cbb839cbe05303d7705fa',
                integratorFee: {
                    share: _limitordersdk.Bps.fromPercent(50),
                    receiver: _limitordersdk.Address.fromBigInt(10n),
                    value: new _limitordersdk.Bps(1n)
                }
            });
        }).toThrow(/replace/);
        expect(function() {
            return _quoterrequest.QuoterRequest.new({
                fromTokenAddress: '0x6b175474e89094c44da98b954eedeac495271d0f',
                toTokenAddress: _limitordersdk.Address.ZERO_ADDRESS.toString(),
                amount: '1000000000000000000000',
                walletAddress: '0x00000000219ab540356cbb839cbe05303d7705fa',
                integratorFee: {
                    share: _limitordersdk.Bps.fromPercent(50),
                    receiver: _limitordersdk.Address.fromBigInt(10n),
                    value: new _limitordersdk.Bps(1n)
                }
            });
        }).toThrow(/replace/);
    });
    it('returns error fromTokenAddress equals toTokenAddress', function() {
        expect(function() {
            return _quoterrequest.QuoterRequest.new({
                fromTokenAddress: '0x6b175474e89094c44da98b954eedeac495271d0f',
                toTokenAddress: '0x6b175474e89094c44da98b954eedeac495271d0f',
                amount: '1000000000000000000000',
                walletAddress: '0x00000000219ab540356cbb839cbe05303d7705fa',
                integratorFee: {
                    share: _limitordersdk.Bps.fromPercent(50),
                    receiver: _limitordersdk.Address.fromBigInt(10n),
                    value: new _limitordersdk.Bps(1n)
                }
            });
        }).toThrow(/fromTokenAddress and toTokenAddress should be different/);
    });
    it('returns error if walletAddress invalid', function() {
        expect(function() {
            return _quoterrequest.QuoterRequest.new({
                fromTokenAddress: '0x6b175474e89094c44da98b954eedeac495271d0f',
                toTokenAddress: '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2',
                amount: '1000000000000000000000',
                walletAddress: '0x0000000019ab540356cbb839be05303d7705fa1',
                integratorFee: {
                    share: _limitordersdk.Bps.fromPercent(50),
                    receiver: _limitordersdk.Address.fromBigInt(10n),
                    value: new _limitordersdk.Bps(1n)
                }
            });
        }).toThrow(/Invalid address/);
    });
    it('returns error if amount is invalid', function() {
        expect(function() {
            return _quoterrequest.QuoterRequest.new({
                fromTokenAddress: '0x6b175474e89094c44da98b954eedeac495271d0f',
                toTokenAddress: '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2',
                amount: 'dasdad',
                walletAddress: '0x00000000219ab540356cbb839cbe05303d7705fa',
                integratorFee: {
                    share: _limitordersdk.Bps.fromPercent(50),
                    receiver: _limitordersdk.Address.fromBigInt(10n),
                    value: new _limitordersdk.Bps(1n)
                }
            });
        }).toThrow(/is invalid amount/);
    });
    it('returns error if  fee is provided and source not', function() {
        expect(function() {
            return _quoterrequest.QuoterRequest.new({
                fromTokenAddress: '0x6b175474e89094c44da98b954eedeac495271d0f',
                toTokenAddress: '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2',
                amount: '1000000',
                walletAddress: '0x00000000219ab540356cbb839cbe05303d7705fa',
                integratorFee: {
                    share: _limitordersdk.Bps.fromPercent(50),
                    receiver: _limitordersdk.Address.fromBigInt(10n),
                    value: new _limitordersdk.Bps(1n)
                }
            });
        }).toThrow(/cannot use fee without source/);
    });
});
