// @1inch/limit-order-sdk/dist/esm/limit-order/interaction.js
function _class_call_check(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}
function _defineProperties(target, props) {
    for(var i = 0; i < props.length; i++){
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
    }
}
function _create_class(Constructor, protoProps, staticProps) {
    if (protoProps) _defineProperties(Constructor.prototype, protoProps);
    if (staticProps) _defineProperties(Constructor, staticProps);
    return Constructor;
}
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
import { BytesIter, isHexBytes, trim0x } from '@1inch/byte-utils';
import assert from 'assert';
import { Address } from '../address.js';
export var Interaction = /*#__PURE__*/ function() {
    "use strict";
    function Interaction(target, data) {
        _class_call_check(this, Interaction);
        _define_property(this, "target", void 0);
        _define_property(this, "data", void 0);
        this.target = target;
        this.data = data;
        assert(isHexBytes(data), 'Interaction data must be valid hex bytes');
    }
    _create_class(Interaction, [
        {
            key: "encode",
            value: /**
     * Hex string with 0x. First 20 bytes are target, then data
     */ function encode() {
                return this.target.toString() + trim0x(this.data);
            }
        }
    ], [
        {
            key: "decode",
            value: /**
     * Create `Interaction` from bytes
     *
     * @param bytes Hex string with 0x. First 20 bytes are target, then data
     */ function decode(bytes) {
                var iter = BytesIter.HexString(bytes);
                return new Interaction(new Address(iter.nextUint160()), iter.rest());
            }
        }
    ]);
    return Interaction;
}();
