// @1inch/fusion-sdk/dist/cjs/api/quoter/quoter.api.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
var _limitordersdk = require("@1inch/limit-order-sdk");
var _quoterapi = require("./quoter.api.js");
var _quoterrequest = require("./quoter.request.js");
var _index = require("./quote/index.js");
var _types = require("./types.js");
var _quotercustompresetrequest = require("./quoter-custom-preset.request.js");
var _constants = require("../../constants.js");
function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) {
    try {
        var info = gen[key](arg);
        var value = info.value;
    } catch (error) {
        reject(error);
        return;
    }
    if (info.done) {
        resolve(value);
    } else {
        Promise.resolve(value).then(_next, _throw);
    }
}
function _async_to_generator(fn) {
    return function() {
        var self = this, args = arguments;
        return new Promise(function(resolve, reject) {
            var gen = fn.apply(self, args);
            function _next(value) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value);
            }
            function _throw(err) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err);
            }
            _next(undefined);
        });
    };
}
function _ts_generator(thisArg, body) {
    var f, y, t, g, _ = {
        label: 0,
        sent: function() {
            if (t[0] & 1) throw t[1];
            return t[1];
        },
        trys: [],
        ops: []
    };
    return g = {
        next: verb(0),
        "throw": verb(1),
        "return": verb(2)
    }, typeof Symbol === "function" && (g[Symbol.iterator] = function() {
        return this;
    }), g;
    function verb(n) {
        return function(v) {
            return step([
                n,
                v
            ]);
        };
    }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while(_)try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [
                op[0] & 2,
                t.value
            ];
            switch(op[0]){
                case 0:
                case 1:
                    t = op;
                    break;
                case 4:
                    _.label++;
                    return {
                        value: op[1],
                        done: false
                    };
                case 5:
                    _.label++;
                    y = op[1];
                    op = [
                        0
                    ];
                    continue;
                case 7:
                    op = _.ops.pop();
                    _.trys.pop();
                    continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                        _ = 0;
                        continue;
                    }
                    if (op[0] === 3 && (!t || op[1] > t[0] && op[1] < t[3])) {
                        _.label = op[1];
                        break;
                    }
                    if (op[0] === 6 && _.label < t[1]) {
                        _.label = t[1];
                        t = op;
                        break;
                    }
                    if (t && _.label < t[2]) {
                        _.label = t[2];
                        _.ops.push(op);
                        break;
                    }
                    if (t[2]) _.ops.pop();
                    _.trys.pop();
                    continue;
            }
            op = body.call(thisArg, _);
        } catch (e) {
            op = [
                6,
                e
            ];
            y = 0;
        } finally{
            f = t = 0;
        }
        if (op[0] & 5) throw op[1];
        return {
            value: op[0] ? op[1] : void 0,
            done: true
        };
    }
}
describe('Quoter API', function() {
    var httpProvider;
    beforeEach(function() {
        httpProvider = {
            get: jest.fn().mockImplementationOnce(function() {
                return Promise.resolve(ResponseMock);
            }),
            post: jest.fn().mockImplementation(function() {
                return Promise.resolve(ResponseMock);
            })
        };
    });
    var params = _quoterrequest.QuoterRequest.new({
        fromTokenAddress: '0x6b175474e89094c44da98b954eedeac495271d0f',
        toTokenAddress: '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2',
        amount: '1000000000000000000000',
        walletAddress: '0x00000000219ab540356cbb839cbe05303d7705fa'
    });
    var ResponseMock = {
        fromTokenAmount: '1000000000000000000000',
        recommended_preset: _types.PresetEnum.medium,
        autoK: 5.5,
        presets: {
            fast: {
                auctionDuration: 180,
                startAuctionIn: 36,
                bankFee: '0',
                initialRateBump: 200461,
                auctionStartAmount: '626771998563995046',
                auctionEndAmount: '614454580595911348',
                tokenFee: '9183588477842300',
                points: [
                    {
                        delay: 24,
                        coefficient: 50461
                    }
                ],
                allowPartialFills: true,
                allowMultipleFills: true,
                exclusiveResolver: null,
                gasCost: {
                    gasBumpEstimate: 0,
                    gasPriceEstimate: '0'
                }
            },
            medium: {
                auctionDuration: 180,
                startAuctionIn: 12,
                bankFee: '0',
                initialRateBump: 210661,
                auctionStartAmount: '627398742236202876',
                auctionEndAmount: '614454580595911348',
                tokenFee: '9183588477842300',
                points: [
                    {
                        delay: 24,
                        coefficient: 50461
                    }
                ],
                allowPartialFills: true,
                allowMultipleFills: true,
                exclusiveResolver: null,
                gasCost: {
                    gasBumpEstimate: 0,
                    gasPriceEstimate: '0'
                }
            },
            slow: {
                auctionDuration: 600,
                startAuctionIn: 12,
                bankFee: '0',
                initialRateBump: 302466,
                auctionStartAmount: '633039742513363640',
                auctionEndAmount: '614454580595911348',
                tokenFee: '9183588477842300',
                points: [
                    {
                        delay: 24,
                        coefficient: 50461
                    }
                ],
                allowPartialFills: true,
                allowMultipleFills: true,
                exclusiveResolver: null,
                gasCost: {
                    gasBumpEstimate: 0,
                    gasPriceEstimate: '0'
                }
            }
        },
        toTokenAmount: '626772029219852913',
        prices: {
            usd: {
                fromToken: '0.99326233048693179928',
                toToken: '1618.25668999999970765202'
            }
        },
        volume: {
            usd: {
                fromToken: '993.26233048693179928',
                toToken: '1014.278029389902274042'
            }
        },
        quoteId: null,
        settlementAddress: '0xa88800cd213da5ae406ce248380802bd53b47647',
        whitelist: [
            '0x84d99aa569d93a9ca187d83734c8c4a519c4e9b1',
            '0xcfa62f77920d6383be12c91c71bd403599e1116f'
        ],
        fee: {
            whitelistDiscountPercent: 1,
            receiver: _constants.ONE_INCH_LIMIT_ORDER_V4,
            bps: 10
        },
        marketAmount: '626772029219852913'
    };
    var QuoterResponseMock = new _index.Quote(params, ResponseMock);
    it('should get quote with disabled estimate', /*#__PURE__*/ _async_to_generator(function() {
        var quoter, res;
        return _ts_generator(this, function(_state) {
            switch(_state.label){
                case 0:
                    quoter = _quoterapi.QuoterApi.new({
                        url: 'https://test.com/quoter',
                        network: 1
                    }, httpProvider);
                    return [
                        4,
                        quoter.getQuote(params)
                    ];
                case 1:
                    res = _state.sent();
                    expect(res).toStrictEqual(QuoterResponseMock);
                    expect(httpProvider.get).toHaveBeenCalledWith('https://test.com/quoter/v2.0/1/quote/receive/?fromTokenAddress=0x6b175474e89094c44da98b954eedeac495271d0f&toTokenAddress=0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2&amount=1000000000000000000000&walletAddress=0x00000000219ab540356cbb839cbe05303d7705fa&source=sdk&surplus=true');
                    return [
                        2
                    ];
            }
        });
    }));
    it('should not throw error with fee and source added', /*#__PURE__*/ _async_to_generator(function() {
        var quoter, params, QuoterResponseMock, res;
        return _ts_generator(this, function(_state) {
            switch(_state.label){
                case 0:
                    quoter = _quoterapi.QuoterApi.new({
                        url: 'https://test.com/quoter',
                        network: 1
                    }, httpProvider);
                    params = _quoterrequest.QuoterRequest.new({
                        fromTokenAddress: '0x6b175474e89094c44da98b954eedeac495271d0f',
                        toTokenAddress: '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2',
                        amount: '1000000000000000000000',
                        walletAddress: '0x00000000219ab540356cbb839cbe05303d7705fa',
                        integratorFee: {
                            share: _limitordersdk.Bps.fromPercent(50),
                            receiver: _limitordersdk.Address.fromBigInt(10n),
                            value: new _limitordersdk.Bps(1n)
                        },
                        source: '0x6b175474e89094c44da98b954eedeac495271d0f'
                    });
                    QuoterResponseMock = new _index.Quote(params, ResponseMock);
                    return [
                        4,
                        quoter.getQuote(params)
                    ];
                case 1:
                    res = _state.sent();
                    expect(res).toStrictEqual(QuoterResponseMock);
                    expect(httpProvider.get).toHaveBeenCalledWith('https://test.com/quoter/v2.0/1/quote/receive/?fromTokenAddress=0x6b175474e89094c44da98b954eedeac495271d0f&toTokenAddress=0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2&amount=1000000000000000000000&walletAddress=0x00000000219ab540356cbb839cbe05303d7705fa&fee=1&source=0x6b175474e89094c44da98b954eedeac495271d0f&surplus=true');
                    return [
                        2
                    ];
            }
        });
    }));
    it('should not throw error with slippage added', /*#__PURE__*/ _async_to_generator(function() {
        var quoter, params, QuoterResponseMock, res;
        return _ts_generator(this, function(_state) {
            switch(_state.label){
                case 0:
                    quoter = _quoterapi.QuoterApi.new({
                        url: 'https://test.com/quoter',
                        network: 1
                    }, httpProvider);
                    params = _quoterrequest.QuoterRequest.new({
                        fromTokenAddress: '0x6b175474e89094c44da98b954eedeac495271d0f',
                        toTokenAddress: '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2',
                        amount: '1000000000000000000000',
                        walletAddress: '0x00000000219ab540356cbb839cbe05303d7705fa',
                        integratorFee: {
                            share: _limitordersdk.Bps.fromPercent(50),
                            receiver: _limitordersdk.Address.fromBigInt(10n),
                            value: new _limitordersdk.Bps(1n)
                        },
                        source: '0x6b175474e89094c44da98b954eedeac495271d0f',
                        slippage: 1
                    });
                    QuoterResponseMock = new _index.Quote(params, ResponseMock);
                    return [
                        4,
                        quoter.getQuote(params)
                    ];
                case 1:
                    res = _state.sent();
                    expect(res).toStrictEqual(QuoterResponseMock);
                    expect(httpProvider.get).toHaveBeenCalledWith('https://test.com/quoter/v2.0/1/quote/receive/?fromTokenAddress=0x6b175474e89094c44da98b954eedeac495271d0f&toTokenAddress=0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2&amount=1000000000000000000000&walletAddress=0x00000000219ab540356cbb839cbe05303d7705fa&fee=1&source=0x6b175474e89094c44da98b954eedeac495271d0f&surplus=true&slippage=1');
                    return [
                        2
                    ];
            }
        });
    }));
    it('getQuoteWithCustomPreset', /*#__PURE__*/ _async_to_generator(function() {
        var quoter, params, body, QuoterResponseMock, res;
        return _ts_generator(this, function(_state) {
            switch(_state.label){
                case 0:
                    quoter = _quoterapi.QuoterApi.new({
                        url: 'https://test.com/quoter',
                        network: 1
                    }, httpProvider);
                    params = _quoterrequest.QuoterRequest.new({
                        fromTokenAddress: '0x6b175474e89094c44da98b954eedeac495271d0f',
                        toTokenAddress: '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2',
                        amount: '1000000000000000000000',
                        walletAddress: '0x00000000219ab540356cbb839cbe05303d7705fa',
                        integratorFee: {
                            share: _limitordersdk.Bps.fromPercent(50),
                            receiver: _limitordersdk.Address.fromBigInt(10n),
                            value: new _limitordersdk.Bps(1n)
                        },
                        source: '0x6b175474e89094c44da98b954eedeac495271d0f'
                    });
                    body = _quotercustompresetrequest.QuoterCustomPresetRequest.new({
                        customPreset: {
                            auctionDuration: 180,
                            auctionStartAmount: '100000',
                            auctionEndAmount: '50000',
                            points: [
                                {
                                    toTokenAmount: '90000',
                                    delay: 20
                                },
                                {
                                    toTokenAmount: '70000',
                                    delay: 40
                                }
                            ]
                        }
                    });
                    QuoterResponseMock = new _index.Quote(params, ResponseMock);
                    return [
                        4,
                        quoter.getQuoteWithCustomPreset(params, body)
                    ];
                case 1:
                    res = _state.sent();
                    expect(res).toStrictEqual(QuoterResponseMock);
                    expect(httpProvider.post).toHaveBeenCalledWith('https://test.com/quoter/v2.0/1/quote/receive/?fromTokenAddress=0x6b175474e89094c44da98b954eedeac495271d0f&toTokenAddress=0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2&amount=1000000000000000000000&walletAddress=0x00000000219ab540356cbb839cbe05303d7705fa&fee=1&source=0x6b175474e89094c44da98b954eedeac495271d0f&surplus=true', body.build());
                    return [
                        2
                    ];
            }
        });
    }));
    it('getQuoteWithCustomPreset with slippage', /*#__PURE__*/ _async_to_generator(function() {
        var quoter, params, body, QuoterResponseMock, res;
        return _ts_generator(this, function(_state) {
            switch(_state.label){
                case 0:
                    quoter = _quoterapi.QuoterApi.new({
                        url: 'https://test.com/quoter',
                        network: 1
                    }, httpProvider);
                    params = _quoterrequest.QuoterRequest.new({
                        fromTokenAddress: '0x6b175474e89094c44da98b954eedeac495271d0f',
                        toTokenAddress: '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2',
                        amount: '1000000000000000000000',
                        walletAddress: '0x00000000219ab540356cbb839cbe05303d7705fa',
                        integratorFee: {
                            share: _limitordersdk.Bps.fromPercent(50),
                            receiver: _limitordersdk.Address.fromBigInt(10n),
                            value: new _limitordersdk.Bps(1n)
                        },
                        source: '0x6b175474e89094c44da98b954eedeac495271d0f',
                        slippage: 1
                    });
                    body = _quotercustompresetrequest.QuoterCustomPresetRequest.new({
                        customPreset: {
                            auctionDuration: 180,
                            auctionStartAmount: '100000',
                            auctionEndAmount: '50000',
                            points: [
                                {
                                    toTokenAmount: '90000',
                                    delay: 20
                                },
                                {
                                    toTokenAmount: '70000',
                                    delay: 40
                                }
                            ]
                        }
                    });
                    QuoterResponseMock = new _index.Quote(params, ResponseMock);
                    return [
                        4,
                        quoter.getQuoteWithCustomPreset(params, body)
                    ];
                case 1:
                    res = _state.sent();
                    expect(res).toStrictEqual(QuoterResponseMock);
                    expect(httpProvider.post).toHaveBeenCalledWith('https://test.com/quoter/v2.0/1/quote/receive/?fromTokenAddress=0x6b175474e89094c44da98b954eedeac495271d0f&toTokenAddress=0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2&amount=1000000000000000000000&walletAddress=0x00000000219ab540356cbb839cbe05303d7705fa&fee=1&source=0x6b175474e89094c44da98b954eedeac495271d0f&surplus=true&slippage=1', body.build());
                    return [
                        2
                    ];
            }
        });
    }));
});
