// @1inch/fusion-sdk/dist/cjs/sdk/sdk.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "FusionSDK", {
    enumerable: true,
    get: function() {
        return FusionSDK;
    }
});
var _limitordersdk = require("@1inch/limit-order-sdk");
var _index = require("./encoders/index.js");
var _index1 = require("../api/index.js");
var _index2 = require("../api/orders/index.js");
function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) {
    try {
        var info = gen[key](arg);
        var value = info.value;
    } catch (error) {
        reject(error);
        return;
    }
    if (info.done) {
        resolve(value);
    } else {
        Promise.resolve(value).then(_next, _throw);
    }
}
function _async_to_generator(fn) {
    return function() {
        var self = this, args = arguments;
        return new Promise(function(resolve, reject) {
            var gen = fn.apply(self, args);
            function _next(value) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value);
            }
            function _throw(err) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err);
            }
            _next(undefined);
        });
    };
}
function _class_call_check(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}
function _defineProperties(target, props) {
    for(var i = 0; i < props.length; i++){
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
    }
}
function _create_class(Constructor, protoProps, staticProps) {
    if (protoProps) _defineProperties(Constructor.prototype, protoProps);
    if (staticProps) _defineProperties(Constructor, staticProps);
    return Constructor;
}
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
function _ts_generator(thisArg, body) {
    var f, y, t, g, _ = {
        label: 0,
        sent: function() {
            if (t[0] & 1) throw t[1];
            return t[1];
        },
        trys: [],
        ops: []
    };
    return g = {
        next: verb(0),
        "throw": verb(1),
        "return": verb(2)
    }, typeof Symbol === "function" && (g[Symbol.iterator] = function() {
        return this;
    }), g;
    function verb(n) {
        return function(v) {
            return step([
                n,
                v
            ]);
        };
    }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while(_)try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [
                op[0] & 2,
                t.value
            ];
            switch(op[0]){
                case 0:
                case 1:
                    t = op;
                    break;
                case 4:
                    _.label++;
                    return {
                        value: op[1],
                        done: false
                    };
                case 5:
                    _.label++;
                    y = op[1];
                    op = [
                        0
                    ];
                    continue;
                case 7:
                    op = _.ops.pop();
                    _.trys.pop();
                    continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                        _ = 0;
                        continue;
                    }
                    if (op[0] === 3 && (!t || op[1] > t[0] && op[1] < t[3])) {
                        _.label = op[1];
                        break;
                    }
                    if (op[0] === 6 && _.label < t[1]) {
                        _.label = t[1];
                        t = op;
                        break;
                    }
                    if (t && _.label < t[2]) {
                        _.label = t[2];
                        _.ops.push(op);
                        break;
                    }
                    if (t[2]) _.ops.pop();
                    _.trys.pop();
                    continue;
            }
            op = body.call(thisArg, _);
        } catch (e) {
            op = [
                6,
                e
            ];
            y = 0;
        } finally{
            f = t = 0;
        }
        if (op[0] & 5) throw op[1];
        return {
            value: op[0] ? op[1] : void 0,
            done: true
        };
    }
}
var FusionSDK = /*#__PURE__*/ function() {
    "use strict";
    function FusionSDK(config) {
        _class_call_check(this, FusionSDK);
        _define_property(this, "config", void 0);
        _define_property(this, "api", void 0);
        this.config = config;
        this.api = _index1.FusionApi.new({
            url: config.url,
            network: config.network,
            httpProvider: config.httpProvider,
            authKey: config.authKey
        });
    }
    _create_class(FusionSDK, [
        {
            key: "getActiveOrders",
            value: function getActiveOrders() {
                var _ref = arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : {}, page = _ref.page, limit = _ref.limit;
                var _this = this;
                return _async_to_generator(function() {
                    var request;
                    return _ts_generator(this, function(_state) {
                        request = _index2.ActiveOrdersRequest.new({
                            page: page,
                            limit: limit
                        });
                        return [
                            2,
                            _this.api.getActiveOrders(request)
                        ];
                    });
                })();
            }
        },
        {
            key: "getOrderStatus",
            value: function getOrderStatus(orderHash) {
                var _this = this;
                return _async_to_generator(function() {
                    var request;
                    return _ts_generator(this, function(_state) {
                        request = _index2.OrderStatusRequest.new({
                            orderHash: orderHash
                        });
                        return [
                            2,
                            _this.api.getOrderStatus(request)
                        ];
                    });
                })();
            }
        },
        {
            key: "getOrdersByMaker",
            value: function getOrdersByMaker(param) {
                var limit = param.limit, page = param.page, address = param.address;
                var _this = this;
                return _async_to_generator(function() {
                    var request;
                    return _ts_generator(this, function(_state) {
                        request = _index2.OrdersByMakerRequest.new({
                            limit: limit,
                            page: page,
                            address: address
                        });
                        return [
                            2,
                            _this.api.getOrdersByMaker(request)
                        ];
                    });
                })();
            }
        },
        {
            key: "getQuote",
            value: function getQuote(params) {
                var _this = this;
                return _async_to_generator(function() {
                    var request;
                    return _ts_generator(this, function(_state) {
                        request = _index1.QuoterRequest.new({
                            fromTokenAddress: params.fromTokenAddress,
                            toTokenAddress: params.toTokenAddress,
                            amount: params.amount,
                            walletAddress: params.walletAddress || _limitordersdk.Address.ZERO_ADDRESS.toString(),
                            permit: params.permit,
                            enableEstimate: !!params.enableEstimate,
                            source: params.source,
                            isPermit2: params.isPermit2,
                            integratorFee: params.integratorFee,
                            slippage: params.slippage
                        });
                        return [
                            2,
                            _this.api.getQuote(request)
                        ];
                    });
                })();
            }
        },
        {
            key: "getQuoteWithCustomPreset",
            value: function getQuoteWithCustomPreset(params, body) {
                var _this = this;
                return _async_to_generator(function() {
                    var paramsRequest, bodyRequest;
                    return _ts_generator(this, function(_state) {
                        paramsRequest = _index1.QuoterRequest.new({
                            fromTokenAddress: params.fromTokenAddress,
                            toTokenAddress: params.toTokenAddress,
                            amount: params.amount,
                            walletAddress: params.walletAddress || _limitordersdk.Address.ZERO_ADDRESS.toString(),
                            permit: params.permit,
                            enableEstimate: !!params.enableEstimate,
                            integratorFee: params === null || params === void 0 ? void 0 : params.integratorFee,
                            source: params.source,
                            isPermit2: params.isPermit2
                        });
                        bodyRequest = _index1.QuoterCustomPresetRequest.new({
                            customPreset: body.customPreset
                        });
                        return [
                            2,
                            _this.api.getQuoteWithCustomPreset(paramsRequest, bodyRequest)
                        ];
                    });
                })();
            }
        },
        {
            key: "createOrder",
            value: function createOrder(params) {
                var _this = this;
                return _async_to_generator(function() {
                    var quote, order, hash;
                    return _ts_generator(this, function(_state) {
                        switch(_state.label){
                            case 0:
                                return [
                                    4,
                                    _this.getQuoteResult(params)
                                ];
                            case 1:
                                quote = _state.sent();
                                if (!quote.quoteId) {
                                    throw new Error('quoter has not returned quoteId');
                                }
                                order = quote.createFusionOrder({
                                    receiver: params.receiver ? new _limitordersdk.Address(params.receiver) : undefined,
                                    preset: params.preset,
                                    nonce: params.nonce,
                                    allowPartialFills: params.allowPartialFills,
                                    allowMultipleFills: params.allowMultipleFills,
                                    orderExpirationDelay: params.orderExpirationDelay,
                                    network: _this.config.network
                                });
                                hash = order.getOrderHash(_this.config.network);
                                return [
                                    2,
                                    {
                                        order: order,
                                        hash: hash,
                                        quoteId: quote.quoteId
                                    }
                                ];
                        }
                    });
                })();
            }
        },
        {
            key: "submitOrder",
            value: function submitOrder(order, quoteId) {
                var _this = this;
                return _async_to_generator(function() {
                    var orderStruct, signature, relayerRequest;
                    return _ts_generator(this, function(_state) {
                        switch(_state.label){
                            case 0:
                                if (!_this.config.blockchainProvider) {
                                    throw new Error('blockchainProvider has not set to config');
                                }
                                orderStruct = order.build();
                                return [
                                    4,
                                    _this.config.blockchainProvider.signTypedData(orderStruct.maker, order.getTypedData(_this.config.network))
                                ];
                            case 1:
                                signature = _state.sent();
                                relayerRequest = _index1.RelayerRequest.new({
                                    order: orderStruct,
                                    signature: signature,
                                    quoteId: quoteId,
                                    extension: order.extension.encode()
                                });
                                return [
                                    4,
                                    _this.api.submitOrder(relayerRequest)
                                ];
                            case 2:
                                _state.sent();
                                return [
                                    2,
                                    {
                                        order: orderStruct,
                                        signature: signature,
                                        quoteId: quoteId,
                                        orderHash: order.getOrderHash(_this.config.network),
                                        extension: relayerRequest.extension
                                    }
                                ];
                        }
                    });
                })();
            }
        },
        {
            key: "placeOrder",
            value: function placeOrder(params) {
                var _this = this;
                return _async_to_generator(function() {
                    var _ref, order, quoteId;
                    return _ts_generator(this, function(_state) {
                        switch(_state.label){
                            case 0:
                                return [
                                    4,
                                    _this.createOrder(params)
                                ];
                            case 1:
                                _ref = _state.sent(), order = _ref.order, quoteId = _ref.quoteId;
                                return [
                                    2,
                                    _this.submitOrder(order, quoteId)
                                ];
                        }
                    });
                })();
            }
        },
        {
            key: "buildCancelOrderCallData",
            value: function buildCancelOrderCallData(orderHash) {
                var _this = this;
                return _async_to_generator(function() {
                    var getOrderRequest, orderData, order;
                    return _ts_generator(this, function(_state) {
                        switch(_state.label){
                            case 0:
                                getOrderRequest = _index2.OrderStatusRequest.new({
                                    orderHash: orderHash
                                });
                                return [
                                    4,
                                    _this.api.getOrderStatus(getOrderRequest)
                                ];
                            case 1:
                                orderData = _state.sent();
                                if (!orderData) {
                                    throw new Error("Can not get order with the specified orderHash ".concat(orderHash));
                                }
                                order = orderData.order;
                                return [
                                    2,
                                    (0, _index.encodeCancelOrder)(orderHash, new _limitordersdk.MakerTraits(BigInt(order.makerTraits)))
                                ];
                        }
                    });
                })();
            }
        },
        {
            key: "signOrder",
            value: function signOrder(order) {
                var _this = this;
                return _async_to_generator(function() {
                    var orderStruct, data;
                    return _ts_generator(this, function(_state) {
                        if (!_this.config.blockchainProvider) {
                            throw new Error('blockchainProvider has not set to config');
                        }
                        orderStruct = order.build();
                        data = order.getTypedData(_this.config.network);
                        return [
                            2,
                            _this.config.blockchainProvider.signTypedData(orderStruct.maker, data)
                        ];
                    });
                })();
            }
        },
        {
            key: "getQuoteResult",
            value: function getQuoteResult(params) {
                var _this = this;
                return _async_to_generator(function() {
                    var quoterRequest, quoterWithCustomPresetBodyRequest;
                    return _ts_generator(this, function(_state) {
                        quoterRequest = _index1.QuoterRequest.new({
                            fromTokenAddress: params.fromTokenAddress,
                            toTokenAddress: params.toTokenAddress,
                            amount: params.amount,
                            walletAddress: params.walletAddress,
                            permit: params.permit,
                            enableEstimate: true,
                            source: params.source,
                            isPermit2: params.isPermit2,
                            integratorFee: params === null || params === void 0 ? void 0 : params.integratorFee,
                            slippage: params === null || params === void 0 ? void 0 : params.slippage
                        });
                        if (!params.customPreset) {
                            return [
                                2,
                                _this.api.getQuote(quoterRequest)
                            ];
                        }
                        quoterWithCustomPresetBodyRequest = _index1.QuoterCustomPresetRequest.new({
                            customPreset: params.customPreset
                        });
                        return [
                            2,
                            _this.api.getQuoteWithCustomPreset(quoterRequest, quoterWithCustomPresetBodyRequest)
                        ];
                    });
                })();
            }
        }
    ]);
    return FusionSDK;
}();
