// @1inch/limit-order-sdk/dist/cjs/limit-order/eip712/domain.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    EIP712Domain: function() {
        return EIP712Domain;
    },
    LimitOrderV4TypeDataName: function() {
        return LimitOrderV4TypeDataName;
    },
    LimitOrderV4TypeDataVersion: function() {
        return LimitOrderV4TypeDataVersion;
    },
    Order: function() {
        return Order;
    }
});
var EIP712Domain = [
    {
        name: 'name',
        type: 'string'
    },
    {
        name: 'version',
        type: 'string'
    },
    {
        name: 'chainId',
        type: 'uint256'
    },
    {
        name: 'verifyingContract',
        type: 'address'
    }
];
var Order = [
    {
        name: 'salt',
        type: 'uint256'
    },
    {
        name: 'maker',
        type: 'address'
    },
    {
        name: 'receiver',
        type: 'address'
    },
    {
        name: 'makerAsset',
        type: 'address'
    },
    {
        name: 'takerAsset',
        type: 'address'
    },
    {
        name: 'makingAmount',
        type: 'uint256'
    },
    {
        name: 'takingAmount',
        type: 'uint256'
    },
    {
        name: 'makerTraits',
        type: 'uint256'
    }
];
var LimitOrderV4TypeDataName = '1inch Aggregation Router';
var LimitOrderV4TypeDataVersion = '6';
