// @1inch/fusion-sdk/dist/esm/constants.js
export var ZX = '0x';
export var NetworkEnum = /*#__PURE__*/ function(NetworkEnum) {
    NetworkEnum[NetworkEnum["ETHEREUM"] = 1] = "ETHEREUM";
    NetworkEnum[NetworkEnum["POLYGON"] = 137] = "POLYGON";
    NetworkEnum[NetworkEnum["ZKSYNC"] = 324] = "ZKSYNC";
    NetworkEnum[NetworkEnum["BINANCE"] = 56] = "BINANCE";
    NetworkEnum[NetworkEnum["ARBITRUM"] = 42161] = "ARBITRUM";
    NetworkEnum[NetworkEnum["AVALANCHE"] = 43114] = "AVALANCHE";
    NetworkEnum[NetworkEnum["OPTIMISM"] = 10] = "OPTIMISM";
    NetworkEnum[NetworkEnum["FANTOM"] = 250] = "FANTOM";
    NetworkEnum[NetworkEnum["GNOSIS"] = 100] = "GNOSIS";
    NetworkEnum[NetworkEnum["COINBASE"] = 8453] = "COINBASE";
    NetworkEnum[NetworkEnum["LINEA"] = 59144] = "LINEA";
    NetworkEnum[NetworkEnum["SONIC"] = 146] = "SONIC";
    NetworkEnum[NetworkEnum["UNICHAIN"] = 130] = "UNICHAIN";
    return NetworkEnum;
}({});
export var ONE_INCH_LIMIT_ORDER_V4 = '0x111111125421ca6dc452d289314280a0f8842a65';
export var UINT_160_MAX = (1n << 160n) - 1n;
export var UINT_16_MAX = (1n << 16n) - 1n;
export var UINT_80_MAX = (1n << 80n) - 1n;
export var UINT_40_MAX = (1n << 40n) - 1n;
export var UINT_32_MAX = (1n << 32n) - 1n;
export var UINT_24_MAX = (1n << 24n) - 1n;
export var UINT_256_MAX = (1n << 256n) - 1n;
