// @1inch/fusion-sdk/dist/cjs/fusion-order/whitelist/whitelist.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Whitelist", {
    enumerable: true,
    get: function() {
        return Whitelist;
    }
});
var _byteutils = require("@1inch/byte-utils");
var _assert = /*#__PURE__*/ _interop_require_default(require("assert"));
var _constants = require("../../constants.js");
var _time = require("../../utils/time.js");
var _utils = require("../../utils.js");
function _class_call_check(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}
function _defineProperties(target, props) {
    for(var i = 0; i < props.length; i++){
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
    }
}
function _create_class(Constructor, protoProps, staticProps) {
    if (protoProps) _defineProperties(Constructor.prototype, protoProps);
    if (staticProps) _defineProperties(Constructor, staticProps);
    return Constructor;
}
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
var Whitelist = /*#__PURE__*/ function() {
    "use strict";
    function Whitelist(resolvingStartTime, whitelist) {
        _class_call_check(this, Whitelist);
        _define_property(this, "resolvingStartTime", void 0);
        _define_property(this, "whitelist", void 0);
        this.resolvingStartTime = resolvingStartTime;
        this.whitelist = whitelist;
        (0, _assert.default)(whitelist.length, 'whitelist can not be empty');
        whitelist.forEach(function(w) {
            (0, _assert.default)(w.delay < _constants.UINT_16_MAX, 'too big diff between timestamps');
            (0, _assert.default)(w.delay >= 0n, 'delay can not be negative');
        });
    }
    _create_class(Whitelist, [
        {
            key: "length",
            get: function get() {
                return this.whitelist.length;
            }
        },
        {
            key: "canExecuteAt",
            value: /**
     * Check whether address allowed to execute order at the given time
     *
     * @param executor address of executor
     * @param executionTime timestamp in sec at which order planning to execute
     */ function canExecuteAt(executor, executionTime) {
                var addressHalf = executor.toString().slice(-20);
                var allowedFrom = this.resolvingStartTime;
                var _iteratorNormalCompletion = true, _didIteratorError = false, _iteratorError = undefined;
                try {
                    for(var _iterator = this.whitelist[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true){
                        var whitelist = _step.value;
                        allowedFrom += whitelist.delay;
                        if (addressHalf === whitelist.addressHalf) {
                            return executionTime >= allowedFrom;
                        } else if (executionTime < allowedFrom) {
                            return false;
                        }
                    }
                } catch (err) {
                    _didIteratorError = true;
                    _iteratorError = err;
                } finally{
                    try {
                        if (!_iteratorNormalCompletion && _iterator.return != null) {
                            _iterator.return();
                        }
                    } finally{
                        if (_didIteratorError) {
                            throw _iteratorError;
                        }
                    }
                }
                return false;
            }
        },
        {
            key: "isExclusivityPeriod",
            value: function isExclusivityPeriod() {
                var time = arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : (0, _time.now)();
                if (this.whitelist.length === 1) {
                    return true;
                }
                if (this.whitelist[0].delay === this.whitelist[1].delay) {
                    return false;
                }
                return time <= this.resolvingStartTime + this.whitelist[1].delay;
            }
        },
        {
            key: "isExclusiveResolver",
            value: function isExclusiveResolver(wallet) {
                var addressHalf = wallet.toString().slice(-20);
                // only one resolver, so check if it is the passed address
                if (this.whitelist.length === 1) {
                    return addressHalf === this.whitelist[0].addressHalf;
                }
                // more than 1 address can fill at the same time, no exclusivity
                if (this.whitelist[0].delay === this.whitelist[1].delay) {
                    return false;
                }
                return addressHalf === this.whitelist[0].addressHalf;
            }
        },
        {
            key: "isWhitelisted",
            value: function isWhitelisted(address) {
                var half = address.lastHalf();
                return this.whitelist.some(function(w) {
                    return w.addressHalf === half;
                });
            }
        },
        {
            key: "encodeInto",
            value: function encodeInto() {
                var builder = arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : new _byteutils.BytesBuilder();
                builder.addUint32(this.resolvingStartTime).addUint8(BigInt(this.whitelist.length));
                var _iteratorNormalCompletion = true, _didIteratorError = false, _iteratorError = undefined;
                try {
                    // whitelist data
                    for(var _iterator = this.whitelist[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true){
                        var wl = _step.value;
                        builder.addBytes((0, _utils.add0x)(wl.addressHalf)).addUint16(wl.delay);
                    }
                } catch (err) {
                    _didIteratorError = true;
                    _iteratorError = err;
                } finally{
                    try {
                        if (!_iteratorNormalCompletion && _iterator.return != null) {
                            _iterator.return();
                        }
                    } finally{
                        if (_didIteratorError) {
                            throw _iteratorError;
                        }
                    }
                }
                return builder;
            }
        },
        {
            key: "encode",
            value: /**
     * Encode whitelist data to 0x prefixed bytes string
     */ function encode() {
                return this.encodeInto().asHex();
            }
        },
        {
            key: "equal",
            value: function equal(other) {
                return this.whitelist.length === other.whitelist.length && this.whitelist.every(function(val, i) {
                    return other.whitelist[i].delay === val.delay && other.whitelist[i].addressHalf === val.addressHalf;
                });
            }
        }
    ], [
        {
            key: "decodeFrom",
            value: /**
     * Construct `Whitelist` from BytesIter
     *
     * @param bytes with 0x prefix in next format:
     * - uint32 auctionStartTime
     * - uint8 whitelist size
     * - (bytes10 last10bytesOfAddress, uint16 auctionDelay) * N whitelist info
     */ function decodeFrom(bytes) {
                var whitelist = [];
                var resolvingStartTime = BigInt(bytes.nextUint32());
                var size = BigInt(bytes.nextUint8());
                for(var i = 0; i < size; i++){
                    var addressHalf = BigInt(bytes.nextBytes(10)).toString(16).padStart(20, '0');
                    var delay = BigInt(bytes.nextUint16());
                    whitelist.push({
                        addressHalf: addressHalf,
                        delay: delay
                    });
                }
                return new Whitelist(resolvingStartTime, whitelist);
            }
        },
        {
            key: "decode",
            value: /**
     * Construct `Whitelist` from bytes
     *
     * @param bytes with 0x prefix
     * @see decodeFrom
     */ function decode(bytes) {
                return Whitelist.decodeFrom(_byteutils.BytesIter.HexString(bytes));
            }
        },
        {
            key: "new",
            value: function _new(resolvingStartTime, whitelist) /**
             * Timestamp in sec at which address can start resolving
             */ {
                var sumDelay = 0n;
                var _whitelist = whitelist.map(function(d) {
                    return {
                        addressHalf: d.address.toString().slice(-20),
                        allowFrom: d.allowFrom < resolvingStartTime ? resolvingStartTime : d.allowFrom
                    };
                }).sort(function(a, b) {
                    return Number(a.allowFrom - b.allowFrom);
                }).// ASC
                map(function(val) {
                    var delay = val.allowFrom - resolvingStartTime - sumDelay;
                    sumDelay += delay;
                    return {
                        delay: delay,
                        addressHalf: val.addressHalf
                    };
                });
                return new Whitelist(resolvingStartTime, _whitelist);
            }
        },
        {
            key: "fromNow",
            value: function fromNow(whitelist) /**
             * Timestamp in sec at which address can start resolving
             */ {
                return Whitelist.new((0, _time.now)(), whitelist);
            }
        }
    ]);
    return Whitelist;
}();
