// @1inch/fusion-sdk/dist/cjs/fusion-order/fusion-extension.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "FusionExtension", {
    enumerable: true,
    get: function() {
        return FusionExtension;
    }
});
var _limitordersdk = require("@1inch/limit-order-sdk");
var _byteutils = require("@1inch/byte-utils");
var _assert = /*#__PURE__*/ _interop_require_default(require("assert"));
var _index = require("./auction-details/index.js");
var _whitelist = require("./whitelist/whitelist.js");
var _surplusparams = require("./surplus-params.js");
var _index1 = require("./fees/index.js");
var _utils = require("../utils.js");
function _class_call_check(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}
function _defineProperties(target, props) {
    for(var i = 0; i < props.length; i++){
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
    }
}
function _create_class(Constructor, protoProps, staticProps) {
    if (protoProps) _defineProperties(Constructor.prototype, protoProps);
    if (staticProps) _defineProperties(Constructor, staticProps);
    return Constructor;
}
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
var FusionExtension = /*#__PURE__*/ function() {
    "use strict";
    function FusionExtension(address, auctionDetails, whitelist, surplus, extra) {
        _class_call_check(this, FusionExtension);
        _define_property(this, "address", void 0);
        _define_property(this, "auctionDetails", void 0);
        _define_property(this, "whitelist", void 0);
        _define_property(this, "surplus", void 0);
        _define_property(this, "extra", void 0);
        this.address = address;
        this.auctionDetails = auctionDetails;
        this.whitelist = whitelist;
        this.surplus = surplus;
        this.extra = extra;
    }
    _create_class(FusionExtension, [
        {
            key: "build",
            value: function build() {
                var _this_extra;
                var amountData = this.buildAmountGetterData(true);
                var builder = new _limitordersdk.ExtensionBuilder().withMakingAmountData(this.address, amountData).withTakingAmountData(this.address, amountData).withPostInteraction(new _limitordersdk.Interaction(this.address, this.buildInteractionData()));
                if ((_this_extra = this.extra) === null || _this_extra === void 0 ? void 0 : _this_extra.makerPermit) {
                    var _this_extra1, _this_extra2;
                    builder.withMakerPermit((_this_extra1 = this.extra) === null || _this_extra1 === void 0 ? void 0 : _this_extra1.makerPermit.target, (_this_extra2 = this.extra) === null || _this_extra2 === void 0 ? void 0 : _this_extra2.makerPermit.data);
                }
                return builder.build();
            }
        },
        {
            key: "buildInteractionData",
            value: /**
     * Build data for `FeeTaker.postInteraction`
     *
     * Data is built of:
     *  `1 byte` - flags:
     *      01 bit `CUSTOM_RECEIVER_FLAG` - set to 1 if order has custom receiver
     *  `20 bytes` — integrator fee recipient
     *  `20 bytes` - protocol fee recipient
     *  `[20 bytes]` — receiver of taking tokens (optional, if not set, maker is used). See `CUSTOM_RECEIVER_FLAG` flag
     *  `bytes` - same as in `buildAmountGetterData`
     *  `32 bytes` - estimated taking amount
     *  `1 byte` - protocol surplus fee (in 1e2)
     * @see buildAmountGetterData
     * @see https://github.com/1inch/limit-order-protocol/blob/22a18f7f20acfec69d4f50ce1880e8e662477710/contracts/extensions/FeeTaker.sol#L114
     */ function buildInteractionData() {
                var _this_extra, _this_extra_fees, _this_extra1, _this_extra_fees1, _this_extra2;
                var customReceiver = ((_this_extra = this.extra) === null || _this_extra === void 0 ? void 0 : _this_extra.customReceiver) || _limitordersdk.Address.ZERO_ADDRESS;
                var flags = new _byteutils.BN(0n).setBit(FusionExtension.CUSTOM_RECEIVER_FLAG_BIT, Boolean(!customReceiver.isZero()));
                var integratorReceiver = ((_this_extra1 = this.extra) === null || _this_extra1 === void 0 ? void 0 : (_this_extra_fees = _this_extra1.fees) === null || _this_extra_fees === void 0 ? void 0 : _this_extra_fees.integrator.integrator) || _limitordersdk.Address.ZERO_ADDRESS;
                var protocolReceiver = ((_this_extra2 = this.extra) === null || _this_extra2 === void 0 ? void 0 : (_this_extra_fees1 = _this_extra2.fees) === null || _this_extra_fees1 === void 0 ? void 0 : _this_extra_fees1.protocol) || _limitordersdk.Address.ZERO_ADDRESS;
                var builder = new _byteutils.BytesBuilder().addUint8(flags).addAddress(integratorReceiver.toString()).addAddress(protocolReceiver.toString());
                if (!customReceiver.isZero()) {
                    builder.addAddress(customReceiver.toString());
                }
                builder.addBytes(this.buildAmountGetterData(false));
                // surplus params
                builder.addUint256(this.surplus.estimatedTakerAmount);
                builder.addUint8(BigInt(this.surplus.protocolFee.toPercent()));
                return builder.asHex();
            }
        },
        {
            key: "buildAmountGetterData",
            value: /**
     * Build data for getMakingAmount/getTakingAmount
     *
     * AuctionDetails
     * 2 bytes — integrator fee percentage (in 1e5)
     * 1 byte - integrator share percentage (in 1e2)
     * 2 bytes — resolver fee percentage (in 1e5)
     * 1 byte - whitelist discount numerator (in 1e2)
     * Whitelist
     *
     * @see https://github.com/1inch/limit-order-settlement/blob/82f0a25c969170f710825ce6aa6920062adbde88/contracts/SimpleSettlement.sol#L34
     */ function buildAmountGetterData(forAmountGetters) {
                var _this_extra_fees, _this_extra, _this_extra_fees1, _this_extra1, _this_extra_fees2, _this_extra2, _this_extra_fees3, _this_extra3;
                var builder = new _byteutils.BytesBuilder();
                if (forAmountGetters) {
                    // auction data required only for `getMakingAmount/getTakingAmount` and not for `postInteraction`
                    this.auctionDetails.encodeInto(builder);
                }
                var integrator = {
                    fee: ((_this_extra = this.extra) === null || _this_extra === void 0 ? void 0 : (_this_extra_fees = _this_extra.fees) === null || _this_extra_fees === void 0 ? void 0 : _this_extra_fees.integrator.fee.toFraction(_index1.Fees.BASE_1E5)) || 0,
                    share: ((_this_extra1 = this.extra) === null || _this_extra1 === void 0 ? void 0 : (_this_extra_fees1 = _this_extra1.fees) === null || _this_extra_fees1 === void 0 ? void 0 : _this_extra_fees1.integrator.share.toFraction(_index1.Fees.BASE_1E2)) || 0
                };
                var resolverFee = ((_this_extra2 = this.extra) === null || _this_extra2 === void 0 ? void 0 : (_this_extra_fees2 = _this_extra2.fees) === null || _this_extra_fees2 === void 0 ? void 0 : _this_extra_fees2.resolver.fee.toFraction(_index1.Fees.BASE_1E5)) || 0;
                var whitelistDiscount = ((_this_extra3 = this.extra) === null || _this_extra3 === void 0 ? void 0 : (_this_extra_fees3 = _this_extra3.fees) === null || _this_extra_fees3 === void 0 ? void 0 : _this_extra_fees3.resolver.whitelistDiscount) || _limitordersdk.Bps.ZERO;
                builder.addUint16(BigInt(integrator.fee)).addUint8(BigInt(integrator.share)).addUint16(BigInt(resolverFee)).addUint8(BigInt(// contract expects discount numerator, but class contain discount
                Number(_index1.Fees.BASE_1E2) - whitelistDiscount.toFraction(_index1.Fees.BASE_1E2)));
                if (forAmountGetters) {
                    // amount getters need only addresses, without delays
                    builder.addUint8(BigInt(this.whitelist.whitelist.length));
                    this.whitelist.whitelist.forEach(function(i) {
                        builder.addBytes((0, _utils.add0x)(i.addressHalf));
                    });
                } else {
                    this.whitelist.encodeInto(builder);
                }
                return builder.asHex();
            }
        },
        {
            key: "getFeesForTaker",
            value: function getFeesForTaker(taker) {
                var _this_extra_fees, _this_extra, _this_extra_fees1, _this_extra1, _this_extra_fees2, _this_extra2;
                var whitelistDiscount = ((_this_extra = this.extra) === null || _this_extra === void 0 ? void 0 : (_this_extra_fees = _this_extra.fees) === null || _this_extra_fees === void 0 ? void 0 : _this_extra_fees.resolver.whitelistDiscount.toFraction(_index1.Fees.BASE_1E2)) || 0;
                var discountNumerator = this.whitelist.isWhitelisted(taker) ? (Number(_index1.Fees.BASE_1E2) - whitelistDiscount) / Number(_index1.Fees.BASE_1E2) : 1;
                var resolverFee = discountNumerator * (((_this_extra1 = this.extra) === null || _this_extra1 === void 0 ? void 0 : (_this_extra_fees1 = _this_extra1.fees) === null || _this_extra_fees1 === void 0 ? void 0 : _this_extra_fees1.resolver.fee.toFraction(_index1.Fees.BASE_1E5)) || 0);
                var resolverFeeBN = BigInt(resolverFee);
                var integratorFeeBN = BigInt(((_this_extra2 = this.extra) === null || _this_extra2 === void 0 ? void 0 : (_this_extra_fees2 = _this_extra2.fees) === null || _this_extra_fees2 === void 0 ? void 0 : _this_extra_fees2.integrator.fee.toFraction(_index1.Fees.BASE_1E5)) || 0);
                return {
                    resolverFee: resolverFeeBN,
                    integratorFee: integratorFeeBN
                };
            }
        },
        {
            key: "getTakingAmountWithFee",
            value: /**
     * Returns takingAmount with fee, but without auction bump
     * @param taker
     * @param orderTakingAmount
     * @private
     */ function getTakingAmountWithFee(taker, orderTakingAmount) {
                var fees = this.getFeesForTaker(taker);
                return (0, _limitordersdk.mulDiv)(orderTakingAmount, _index1.Fees.BASE_1E5 + fees.resolverFee + fees.integratorFee, _index1.Fees.BASE_1E5, _limitordersdk.Rounding.Ceil);
            }
        }
    ], [
        {
            key: "decode",
            value: /**
     * Create `FusionExtension` from bytes
     *
     * @param bytes 0x prefixed bytes
     */ function decode(bytes) {
                var extension = _limitordersdk.Extension.decode(bytes);
                return FusionExtension.fromExtension(extension);
            }
        },
        {
            key: "fromExtension",
            value: /**
     * Create `FusionExtension` from `Extension`
     */ function fromExtension(extension) {
                var settlementContract = _limitordersdk.Address.fromFirstBytes(extension.makingAmountData);
                (0, _assert.default)(_limitordersdk.Address.fromFirstBytes(extension.takingAmountData).equal(settlementContract) && _limitordersdk.Address.fromFirstBytes(extension.postInteraction).equal(settlementContract), 'Invalid extension, all calls should be to the same address');
                (0, _assert.default)(extension.takingAmountData == extension.makingAmountData, 'Invalid extension, taking amount data must be equal to making amount data');
                // region Parse postInteraction data
                var interactionBytes = _byteutils.BytesIter.HexString(extension.postInteraction);
                interactionBytes.nextUint160();
                // skip address of extension
                var flags = _byteutils.BN.fromHex(interactionBytes.nextUint8());
                var integratorFeeRecipient = new _limitordersdk.Address(interactionBytes.nextUint160());
                var protocolFeeRecipient = new _limitordersdk.Address(interactionBytes.nextUint160());
                var customReceiver = flags.getBit(FusionExtension.CUSTOM_RECEIVER_FLAG_BIT) ? new _limitordersdk.Address(interactionBytes.nextUint160()) : undefined;
                var interactionData = parseAmountData(interactionBytes);
                var whitelist = _whitelist.Whitelist.decodeFrom(interactionBytes);
                var surplusParams = _surplusparams.SurplusParams.decodeFrom(interactionBytes);
                //endregion Parse postInteraction data
                //region Parse amount data
                var amountBytes = _byteutils.BytesIter.HexString(extension.makingAmountData);
                amountBytes.nextUint160();
                // skip address of extension
                var auctionDetails = _index.AuctionDetails.decodeFrom(amountBytes);
                var amountData = parseAmountData(amountBytes);
                var whitelistAddressLength = Number(amountBytes.nextUint8());
                (0, _assert.default)(whitelist.length === whitelistAddressLength, 'whitelist addresses must be same in interaction data and in amount data');
                var whitelistAddressesFromAmount = [];
                for(var i = 0; i < whitelistAddressLength; i++){
                    whitelistAddressesFromAmount.push(BigInt(amountBytes.nextBytes(10)).toString(16).padStart(20, '0'));
                }
                //endregion Parse amount data
                var makerPermit = extension.hasMakerPermit ? _limitordersdk.Interaction.decode(extension.makerPermit) : undefined;
                (0, _assert.default)(amountData.fees.integratorFee.value === interactionData.fees.integratorFee.value, "invalid extension: integrator fee must be same in interaction data and in amount data");
                (0, _assert.default)(amountData.fees.resolverFee.value === interactionData.fees.resolverFee.value, "invalid extension: resolver fee must be same in interaction data and in amount data");
                (0, _assert.default)(amountData.fees.whitelistDiscount.equal(interactionData.fees.whitelistDiscount), "invalid extension: whitelistDiscount must be same in interaction data and in amount data");
                (0, _assert.default)(amountData.fees.integratorShare.value === interactionData.fees.integratorShare.value, "invalid extension: integrator share must be same in interaction data and in amount data");
                (0, _assert.default)(whitelist.whitelist.every(function(param, i) {
                    var addressHalf = param.addressHalf;
                    return whitelistAddressesFromAmount[i] === addressHalf;
                }), 'whitelist addresses must be same in interaction data and in amount data');
                var hasFees = !integratorFeeRecipient.isZero() || !protocolFeeRecipient.isZero();
                if (!hasFees) {
                    return new FusionExtension(settlementContract, auctionDetails, whitelist, surplusParams, {
                        makerPermit: makerPermit,
                        customReceiver: customReceiver,
                        fees: undefined
                    });
                }
                var fees = new _index1.Fees(new _index1.ResolverFee(protocolFeeRecipient, interactionData.fees.resolverFee, interactionData.fees.whitelistDiscount), interactionData.fees.integratorFee.isZero() ? _index1.IntegratorFee.ZERO : new _index1.IntegratorFee(integratorFeeRecipient, protocolFeeRecipient, interactionData.fees.integratorFee, interactionData.fees.integratorShare));
                return new FusionExtension(settlementContract, auctionDetails, whitelist, surplusParams, {
                    makerPermit: makerPermit,
                    fees: fees,
                    customReceiver: customReceiver
                });
            }
        }
    ]);
    return FusionExtension;
}();
/**
     * Flags for post-interaction data
     * @private
     */ _define_property(FusionExtension, "CUSTOM_RECEIVER_FLAG_BIT", 0n);
function parseAmountData(iter) {
    var fees = {
        integratorFee: _limitordersdk.Bps.fromFraction(Number(iter.nextUint16()), _index1.Fees.BASE_1E5),
        integratorShare: _limitordersdk.Bps.fromFraction(Number(iter.nextUint8()), _index1.Fees.BASE_1E2),
        resolverFee: _limitordersdk.Bps.fromFraction(Number(iter.nextUint16()), _index1.Fees.BASE_1E5),
        whitelistDiscount: _limitordersdk.Bps.fromFraction(Number(_index1.Fees.BASE_1E2) - Number(iter.nextUint8()), // contract uses 1 - discount
        _index1.Fees.BASE_1E2)
    };
    return {
        fees: fees
    };
}
