// @1inch/fusion-sdk/dist/esm/api/quoter/quoter.request.js
function _class_call_check(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}
function _defineProperties(target, props) {
    for(var i = 0; i < props.length; i++){
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
    }
}
function _create_class(Constructor, protoProps, staticProps) {
    if (protoProps) _defineProperties(Constructor.prototype, protoProps);
    if (staticProps) _defineProperties(Constructor, staticProps);
    return Constructor;
}
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
import { Address } from '@1inch/limit-order-sdk';
import { isValidAmount } from '../../validations.js';
export var QuoterRequest = /*#__PURE__*/ function() {
    "use strict";
    function QuoterRequest(params) {
        _class_call_check(this, QuoterRequest);
        _define_property(this, "fromTokenAddress", void 0);
        _define_property(this, "toTokenAddress", void 0);
        _define_property(this, "amount", void 0);
        _define_property(this, "walletAddress", void 0);
        _define_property(this, "enableEstimate", void 0);
        _define_property(this, "permit", void 0);
        _define_property(this, "integratorFee", void 0);
        _define_property(this, "source", void 0);
        _define_property(this, "isPermit2", void 0);
        _define_property(this, "slippage", void 0);
        this.fromTokenAddress = new Address(params.fromTokenAddress);
        this.toTokenAddress = new Address(params.toTokenAddress);
        this.amount = params.amount;
        this.walletAddress = new Address(params.walletAddress);
        this.enableEstimate = params.enableEstimate || false;
        this.permit = params.permit;
        this.integratorFee = params.integratorFee;
        this.source = params.source || 'sdk';
        var _params_isPermit2;
        this.isPermit2 = (_params_isPermit2 = params.isPermit2) !== null && _params_isPermit2 !== void 0 ? _params_isPermit2 : false;
        this.slippage = params.slippage;
        if (this.fromTokenAddress.isNative()) {
            throw new Error("cannot swap ".concat(Address.NATIVE_CURRENCY, ": wrap native currency to it's wrapper fist"));
        }
        if (this.fromTokenAddress.isZero() || this.toTokenAddress.isZero()) {
            throw new Error("replace ".concat(Address.ZERO_ADDRESS, " with ").concat(Address.NATIVE_CURRENCY));
        }
        if (this.fromTokenAddress.equal(this.toTokenAddress)) {
            throw new Error('fromTokenAddress and toTokenAddress should be different');
        }
        if (!isValidAmount(this.amount)) {
            throw new Error("".concat(this.amount, " is invalid amount"));
        }
        if (this.integratorFee && this.source === 'sdk') {
            throw new Error('cannot use fee without source');
        }
    }
    _create_class(QuoterRequest, [
        {
            key: "build",
            value: function build() {
                var _this_integratorFee;
                return {
                    fromTokenAddress: this.fromTokenAddress.toString(),
                    toTokenAddress: this.toTokenAddress.toString(),
                    amount: this.amount,
                    walletAddress: this.walletAddress.toString(),
                    enableEstimate: this.enableEstimate,
                    permit: this.permit,
                    fee: Number(((_this_integratorFee = this.integratorFee) === null || _this_integratorFee === void 0 ? void 0 : _this_integratorFee.value.value) || 0),
                    source: this.source,
                    isPermit2: this.isPermit2,
                    surplus: true,
                    slippage: this.slippage
                };
            }
        }
    ], [
        {
            key: "new",
            value: function _new(params) {
                return new QuoterRequest(params);
            }
        }
    ]);
    return QuoterRequest;
}();
