// @1inch/fusion-sdk/dist/cjs/amount-calculator/amount-calculator.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "AmountCalculator", {
    enumerable: true,
    get: function() {
        return AmountCalculator;
    }
});
var _limitordersdk = require("@1inch/limit-order-sdk");
var _index = require("./auction-calculator/index.js");
var _index1 = require("../fusion-order/index.js");
var _index2 = require("../fusion-order/fees/index.js");
function _class_call_check(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}
function _defineProperties(target, props) {
    for(var i = 0; i < props.length; i++){
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
    }
}
function _create_class(Constructor, protoProps, staticProps) {
    if (protoProps) _defineProperties(Constructor.prototype, protoProps);
    if (staticProps) _defineProperties(Constructor, staticProps);
    return Constructor;
}
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
/**
 * Calculates fees/amount with accounting to auction
 *
 * @see FusionOrder
 */ var AmountCalculator = /*#__PURE__*/ function() {
    "use strict";
    function AmountCalculator(auctionCalculator, feeCalculator) {
        var surplus = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : _index1.SurplusParams.NO_FEE;
        _class_call_check(this, AmountCalculator);
        _define_property(this, "auctionCalculator", void 0);
        _define_property(this, "feeCalculator", void 0);
        _define_property(this, "surplus", void 0);
        this.auctionCalculator = auctionCalculator;
        this.feeCalculator = feeCalculator;
        this.surplus = surplus;
    }
    _create_class(AmountCalculator, [
        {
            key: "getRequiredTakingAmount",
            value: /**
     * Returns adjusted taking amount with included fees and auction bump
     *
     * @param taker address which fill order
     * @param takingAmount base taking amount without auction and fee
     * @param time block time at which order will be filled
     * @param blockBaseFee base fee of block at which order will be filled
     */ function getRequiredTakingAmount(taker, takingAmount, time) {
                var blockBaseFee = arguments.length > 3 && arguments[3] !== void 0 ? arguments[3] : 0n;
                var _this_feeCalculator;
                var _this_feeCalculator_getTakingAmount;
                var withFee = (_this_feeCalculator_getTakingAmount = (_this_feeCalculator = this.feeCalculator) === null || _this_feeCalculator === void 0 ? void 0 : _this_feeCalculator.getTakingAmount(taker, takingAmount)) !== null && _this_feeCalculator_getTakingAmount !== void 0 ? _this_feeCalculator_getTakingAmount : takingAmount;
                return this.getAuctionBumpedAmount(withFee, time, blockBaseFee);
            }
        },
        {
            key: "getRequiredMakingAmount",
            value: /**
     * Returns adjusted making amount with accounting of fees and auction bump
     *
     * @param taker address which fill order
     * @param makingAmount base making amount without auction and fee
     * @param time block time at which order will be filled
     * @param blockBaseFee base fee of block at which order will be filled
     */ function getRequiredMakingAmount(taker, makingAmount, time) {
                var blockBaseFee = arguments.length > 3 && arguments[3] !== void 0 ? arguments[3] : 0n;
                var _this_feeCalculator;
                var _this_feeCalculator_getMakingAmount;
                var withFee = (_this_feeCalculator_getMakingAmount = (_this_feeCalculator = this.feeCalculator) === null || _this_feeCalculator === void 0 ? void 0 : _this_feeCalculator.getMakingAmount(taker, makingAmount)) !== null && _this_feeCalculator_getMakingAmount !== void 0 ? _this_feeCalculator_getMakingAmount : makingAmount;
                var rateBump = this.auctionCalculator.calcRateBump(time, blockBaseFee);
                return _index.AuctionCalculator.calcAuctionMakingAmount(withFee, rateBump);
            }
        },
        {
            key: "getTotalFee",
            value: /**
     * Returns total fee = integrator + protocol
     *
     * @param taker
     * @param takingAmount base taking amount without auction and fee
     * @param time block time at which order will be filled
     * @param blockBaseFee base fee of block at which order will be filled
     */ function getTotalFee(taker, takingAmount, time) {
                var blockBaseFee = arguments.length > 3 && arguments[3] !== void 0 ? arguments[3] : 0n;
                return this.getIntegratorFee(taker, takingAmount, time, blockBaseFee) + this.getProtocolFee(taker, takingAmount, time, blockBaseFee);
            }
        },
        {
            key: "getUserTakingAmount",
            value: /**
     * Returns amount which will receive user
     *
     * @param taker
     * @param makingAmount amount to be filled
     * @param takingAmount base taking amount without auction and fee
     * @param orderMakingAmount full order making amount
     * @param time block time at which order will be filled
     * @param blockBaseFee base fee of block at which order will be filled
     */ function getUserTakingAmount(taker, makingAmount, takingAmount, orderMakingAmount, time) {
                var blockBaseFee = arguments.length > 5 && arguments[5] !== void 0 ? arguments[5] : 0n;
                var whole = this.getRequiredTakingAmount(taker, takingAmount, time, blockBaseFee);
                var preSurplus = whole - this.getTotalFee(taker, takingAmount, time, blockBaseFee);
                var surplusFee = this._getSurplusFee(preSurplus, makingAmount, orderMakingAmount);
                return preSurplus - surplusFee;
            }
        },
        {
            key: "getSurplusFee",
            value: /**
     * Returns amount in taker asset which sent to protocol as part of surplus share
     *
     * @param taker
     * @param makingAmount amount to be filled
     * @param takingAmount base taking amount without auction and fee
     * @param orderMakingAmount full order making amount
     * @param time block time at which order will be filled
     * @param blockBaseFee base fee of block at which order will be filled
     */ function getSurplusFee(taker, makingAmount, takingAmount, orderMakingAmount, time) {
                var blockBaseFee = arguments.length > 5 && arguments[5] !== void 0 ? arguments[5] : 0n;
                var whole = this.getRequiredTakingAmount(taker, takingAmount, time, blockBaseFee);
                var preSurplus = whole - this.getTotalFee(taker, takingAmount, time, blockBaseFee);
                return this._getSurplusFee(preSurplus, makingAmount, orderMakingAmount);
            }
        },
        {
            key: "getResolverFee",
            value: /**
     * Fee in `takerAsset` which resolver pays to resolver fee receiver
     *
     * @param taker who will fill order
     * @param takingAmount taking amount to calculate fee from, must be without fees/auction adjustments
     * @param time block time at which order will be filled
     * @param blockBaseFee base fee of block at which order will be filled
     */ function getResolverFee(taker, takingAmount, time) {
                var blockBaseFee = arguments.length > 3 && arguments[3] !== void 0 ? arguments[3] : 0n;
                var _this_feeCalculator;
                var _this_feeCalculator_getResolverFee;
                return (_this_feeCalculator_getResolverFee = (_this_feeCalculator = this.feeCalculator) === null || _this_feeCalculator === void 0 ? void 0 : _this_feeCalculator.getResolverFee(taker, this.getAuctionBumpedAmount(takingAmount, time, blockBaseFee))) !== null && _this_feeCalculator_getResolverFee !== void 0 ? _this_feeCalculator_getResolverFee : 0n;
            }
        },
        {
            key: "getIntegratorFee",
            value: /**
     * Fee in `takerAsset` which integrator gets to integrator wallet
     *
     * @param taker who will fill order
     * @param takingAmount taking amount to calculate fee from, must be without fees/auction adjustments
     * @param time block time at which order will be filled
     * @param blockBaseFee base fee of block at which order will be filled
     */ function getIntegratorFee(taker, takingAmount, time) {
                var blockBaseFee = arguments.length > 3 && arguments[3] !== void 0 ? arguments[3] : 0n;
                var _this_feeCalculator;
                var _this_feeCalculator_getIntegratorFee;
                return (_this_feeCalculator_getIntegratorFee = (_this_feeCalculator = this.feeCalculator) === null || _this_feeCalculator === void 0 ? void 0 : _this_feeCalculator.getIntegratorFee(taker, this.getAuctionBumpedAmount(takingAmount, time, blockBaseFee))) !== null && _this_feeCalculator_getIntegratorFee !== void 0 ? _this_feeCalculator_getIntegratorFee : 0n;
            }
        },
        {
            key: "getProtocolShareOfIntegratorFee",
            value: /**
     * Fee in `takerAsset` which protocol gets as share from integrator fee
     *
     * @param taker who will fill order
     * @param takingAmount taking amount to calculate fee from, must be without fees/auction adjustments
     * @param time block time at which order will be filled
     * @param blockBaseFee base fee of block at which order will be filled
     */ function getProtocolShareOfIntegratorFee(taker, takingAmount, time) {
                var blockBaseFee = arguments.length > 3 && arguments[3] !== void 0 ? arguments[3] : 0n;
                var _this_feeCalculator;
                var _this_feeCalculator_getProtocolShareOfIntegratorFee;
                return (_this_feeCalculator_getProtocolShareOfIntegratorFee = (_this_feeCalculator = this.feeCalculator) === null || _this_feeCalculator === void 0 ? void 0 : _this_feeCalculator.getProtocolShareOfIntegratorFee(taker, this.getAuctionBumpedAmount(takingAmount, time, blockBaseFee))) !== null && _this_feeCalculator_getProtocolShareOfIntegratorFee !== void 0 ? _this_feeCalculator_getProtocolShareOfIntegratorFee : 0n;
            }
        },
        {
            key: "getProtocolFee",
            value: /**
     * Fee in `takerAsset` which protocol gets
     * It equals to `share from integrator fee plus resolver fee`
     *
     * @param taker who will fill order
     * @param takingAmount taking amount to calculate fee from, must be without fees/auction adjustments
     * @param time block time at which order will be filled
     * @param blockBaseFee base fee of block at which order will be filled
     */ function getProtocolFee(taker, takingAmount, time) {
                var blockBaseFee = arguments.length > 3 && arguments[3] !== void 0 ? arguments[3] : 0n;
                var _this_feeCalculator;
                var _this_feeCalculator_getProtocolFee;
                return (_this_feeCalculator_getProtocolFee = (_this_feeCalculator = this.feeCalculator) === null || _this_feeCalculator === void 0 ? void 0 : _this_feeCalculator.getProtocolFee(taker, this.getAuctionBumpedAmount(takingAmount, time, blockBaseFee))) !== null && _this_feeCalculator_getProtocolFee !== void 0 ? _this_feeCalculator_getProtocolFee : 0n;
            }
        },
        {
            key: "_getSurplusFee",
            value: /**
     * Calculates surplus fee. It will be sent to the same address as `protocolFee`
     *
     * @param userTakingAmount how much user would receive without surplus
     * @param makingAmount making amount to be filled
     * @param orderMakingAmount full order making amount
     */ function _getSurplusFee(userTakingAmount, makingAmount, orderMakingAmount) {
                var estimatedTakingAmount = (0, _limitordersdk.mulDiv)(this.surplus.estimatedTakerAmount, makingAmount, orderMakingAmount);
                if (userTakingAmount > estimatedTakingAmount) {
                    var surplusFee = (userTakingAmount - estimatedTakingAmount) * BigInt(this.surplus.protocolFee.toPercent()) / 100n;
                    return surplusFee;
                }
                return 0n;
            }
        },
        {
            key: "getAuctionBumpedAmount",
            value: function getAuctionBumpedAmount(takingAmount, time) {
                var blockBaseFee = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : 0n;
                var rateBump = this.auctionCalculator.calcRateBump(time, blockBaseFee);
                return _index.AuctionCalculator.calcAuctionTakingAmount(takingAmount, rateBump);
            }
        }
    ], [
        {
            key: "fromExtension",
            value: function fromExtension(ext) {
                var _ext_extra, _ext_extra1;
                return new AmountCalculator(_index.AuctionCalculator.fromAuctionData(ext.auctionDetails), ((_ext_extra = ext.extra) === null || _ext_extra === void 0 ? void 0 : _ext_extra.fees) ? new _limitordersdk.FeeTakerExt.FeeCalculator((_ext_extra1 = ext.extra) === null || _ext_extra1 === void 0 ? void 0 : _ext_extra1.fees, ext.whitelist) : undefined, ext.surplus);
            }
        },
        {
            key: "calcAuctionTakingAmount",
            value: /**
     * Returns amount with applied rate bump and fees
     *
     * @param baseTakingAmount base amount to apply bump to
     * @param rate auction rate bump
     * @param fee all fees applied to amount
     *
     * @see AuctionCalculator.calcInitialRateBump
     */ function calcAuctionTakingAmount(baseTakingAmount, rate) {
                var fee = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : _limitordersdk.Bps.ZERO;
                var withoutFee = _index.AuctionCalculator.calcAuctionTakingAmount(baseTakingAmount, rate);
                if (fee.isZero()) {
                    return withoutFee;
                }
                var numerator = _index2.Fees.BASE_1E5 + BigInt(fee.toFraction(_index2.Fees.BASE_1E5));
                return withoutFee * numerator / _index2.Fees.BASE_1E5;
            }
        },
        {
            key: "extractFeeAmount",
            value: /**
     * Return fee amount in taker asset which is included in `requiredTakingAmount`
     *
     * @param requiredTakingAmount must already contain fee
     * @param fee to extract
     */ function extractFeeAmount(requiredTakingAmount, fee) {
                return requiredTakingAmount - (0, _limitordersdk.mulDiv)(requiredTakingAmount, _index2.Fees.BASE_1E5, _index2.Fees.BASE_1E5 + BigInt(fee.toFraction(_index2.Fees.BASE_1E5)), _limitordersdk.Rounding.Ceil);
            }
        }
    ]);
    return AmountCalculator;
}();
