// @1inch/limit-order-sdk/dist/cjs/limit-order/limit-order.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
var _limitorder = require("./limit-order.js");
var _extensionbuilder = require("./extensions/extension-builder.js");
var _makertraits = require("./maker-traits.js");
var _address = require("../address.js");
describe('Limit Order', function() {
    jest.spyOn(Math, 'random').mockReturnValue(1);
    jest.spyOn(Date, 'now').mockReturnValue(1673549418040);
    it('should create limit order', function() {
        var order = new _limitorder.LimitOrder({
            makerAsset: new _address.Address('0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2'),
            takerAsset: new _address.Address('0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48'),
            makingAmount: 1000000000000000000n,
            takingAmount: 1420000000n,
            maker: new _address.Address('0x00000000219ab540356cbb839cbe05303d7705fa')
        });
        expect(_limitorder.LimitOrder.fromCalldata(order.toCalldata())).toEqual(order);
    });
    it('should create limit order with passed salt', function() {
        var order = new _limitorder.LimitOrder({
            makerAsset: new _address.Address('0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2'),
            takerAsset: new _address.Address('0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48'),
            makingAmount: 1000000000000000000n,
            takingAmount: 1420000000n,
            maker: new _address.Address('0x00000000219ab540356cbb839cbe05303d7705fa'),
            salt: 10n
        });
        expect(_limitorder.LimitOrder.fromCalldata(order.toCalldata())).toEqual(order);
    });
    it('should create limit order with extension and salt', function() {
        var ext = new _extensionbuilder.ExtensionBuilder().withCustomData('0xdeadbeef').build();
        var order = new _limitorder.LimitOrder({
            makerAsset: new _address.Address('0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2'),
            takerAsset: new _address.Address('0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48'),
            makingAmount: 1000000000000000000n,
            takingAmount: 1420000000n,
            maker: new _address.Address('0x00000000219ab540356cbb839cbe05303d7705fa'),
            salt: _limitorder.LimitOrder.buildSalt(ext)
        }, _makertraits.MakerTraits.default(), ext);
        expect(_limitorder.LimitOrder.fromDataAndExtension(order.build(), ext)).toEqual(order);
    });
});
