// @1inch/limit-order-sdk/dist/cjs/limit-order/extensions/fee-taker/fees.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Fees", {
    enumerable: true,
    get: function() {
        return Fees;
    }
});
var _assert = /*#__PURE__*/ _interop_require_default(require("assert"));
var _resolverfee = require("./resolver-fee.js");
var _integratorfee = require("./integrator-fee.js");
function _class_call_check(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}
function _defineProperties(target, props) {
    for(var i = 0; i < props.length; i++){
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
    }
}
function _create_class(Constructor, protoProps, staticProps) {
    if (protoProps) _defineProperties(Constructor.prototype, protoProps);
    if (staticProps) _defineProperties(Constructor, staticProps);
    return Constructor;
}
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
var Fees = /*#__PURE__*/ function() {
    "use strict";
    function Fees(resolver, integrator) {
        _class_call_check(this, Fees);
        _define_property(this, "resolver", void 0);
        _define_property(this, "integrator", void 0);
        this.resolver = resolver;
        this.integrator = integrator;
        if (!resolver.fee.isZero() && !integrator.fee.isZero()) {
            (0, _assert.default)(resolver.receiver.equal(integrator.protocol), 'resolver fee receiver address and integrator fee protocol address must be same');
        }
        (0, _assert.default)(!(resolver.fee.isZero() && integrator.fee.isZero()), 'at least one fee must be set');
        (0, _assert.default)(this.integrator.fee.toFraction() < 0.6553, 'max fee is 65.53%');
        // 2bytes
        (0, _assert.default)(this.resolver.fee.toFraction() < 0.6553, 'max fee is 65.53%');
    }
    _create_class(Fees, [
        {
            key: "protocol",
            get: // 2bytes
            function get() {
                return this.integrator.fee.isZero() ? this.resolver.receiver : this.integrator.protocol;
            }
        }
    ], [
        {
            key: "resolverFee",
            value: function resolverFee(fee) {
                return new Fees(fee, _integratorfee.IntegratorFee.ZERO);
            }
        },
        {
            key: "integratorFee",
            value: function integratorFee(fee) {
                return new Fees(_resolverfee.ResolverFee.ZERO, fee);
            }
        }
    ]);
    return Fees;
}();
/**
     * 100% = 100000
     */ _define_property(Fees, "BASE_1E5", 100000n);
/**
     * 100% = 100
     */ _define_property(Fees, "BASE_1E2", 100n);
