// @1inch/limit-order-sdk/dist/cjs/limit-order/extensions/fee-taker/integrator-fee.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "IntegratorFee", {
    enumerable: true,
    get: function() {
        return IntegratorFee;
    }
});
var _address = require("../../../address.js");
var _bps = require("../../../bps.js");
function _class_call_check(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
/**
 * Integrator gets `share` of `fee` to `integrator` and the rest goes to `protocol`
 */ var IntegratorFee = function IntegratorFee(integrator, protocol, fee, share) {
    "use strict";
    _class_call_check(this, IntegratorFee);
    _define_property(this, "integrator", void 0);
    _define_property(this, "protocol", void 0);
    _define_property(this, "fee", void 0);
    _define_property(this, "share", void 0);
    this.integrator = integrator;
    this.protocol = protocol;
    this.fee = fee;
    this.share = share;
    if (fee.isZero()) {
        if (!share.isZero()) {
            throw new Error('integrator share must be zero if fee is zero');
        }
        if (!integrator.isZero()) {
            throw new Error('integrator address must be zero if fee is zero');
        }
        if (!protocol.isZero()) {
            throw new Error('protocol address must be zero if fee is zero');
        }
    }
    if ((integrator.isZero() || protocol.isZero()) && !fee.isZero()) {
        throw new Error('fee must be zero if integrator or protocol is zero address');
    }
};
_define_property(IntegratorFee, "ZERO", new IntegratorFee(_address.Address.ZERO_ADDRESS, _address.Address.ZERO_ADDRESS, _bps.Bps.ZERO, _bps.Bps.ZERO));
