// @1inch/fusion-sdk/dist/cjs/amount-calculator/amount-calculator.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
var _limitordersdk = require("@1inch/limit-order-sdk");
var _amountcalculator = require("./amount-calculator.js");
var _index = require("./auction-calculator/index.js");
var _index1 = require("../fusion-order/index.js");
var _time = require("../utils/time.js");
var _index2 = require("../fusion-order/fees/index.js");
describe('AmountCalculator', function() {
    it('should correct extract fee', function() {
        var integratorFeeBps = new _limitordersdk.Bps(6n);
        // 60% from 10bps
        var calculator = new _amountcalculator.AmountCalculator(new _index.AuctionCalculator(1738650250n, 180n, 1218519n, [
            {
                coefficient: 609353,
                delay: 180
            }
        ], {
            gasBumpEstimate: 609353n,
            gasPriceEstimate: 1526n
        }), new _limitordersdk.FeeTakerExt.FeeCalculator(_index2.Fees.integratorFee(new _index2.IntegratorFee(new _limitordersdk.Address('0x8e097e5e0493de033270a01b324caf31f464dc67'), new _limitordersdk.Address('0x90cbe4bdd538d6e9b379bff5fe72c3d67a521de5'), new _limitordersdk.Bps(10n), new _limitordersdk.Bps(6000n))), _index1.Whitelist.new(1738650226n, [
            {
                address: _limitordersdk.Address.fromBigInt(1n),
                allowFrom: 0n
            }
        ])));
        var takingAmount = 100000n;
        var requiredTakingAmount = calculator.getRequiredTakingAmount(_limitordersdk.Address.ZERO_ADDRESS, takingAmount, (0, _time.now)(), 10n);
        var integratorFee = calculator.getIntegratorFee(_limitordersdk.Address.ZERO_ADDRESS, takingAmount, (0, _time.now)(), 10n);
        expect(_amountcalculator.AmountCalculator.extractFeeAmount(requiredTakingAmount, integratorFeeBps)).toEqual(integratorFee);
    });
    it('should apply surplus', function() {
        var startTime = 1738650250n;
        var execTime = startTime + 180n;
        // end of auction
        var auction = new _index.AuctionCalculator(startTime, 180n, 1218519n, [
            {
                coefficient: 609353,
                delay: 180
            }
        ], {
            gasBumpEstimate: 609353n,
            gasPriceEstimate: 1526n
        });
        var taker = _limitordersdk.Address.fromBigInt(1n);
        var feeCalculator = new _limitordersdk.FeeTakerExt.FeeCalculator(_index2.Fees.integratorFee(new _index2.IntegratorFee(new _limitordersdk.Address('0x8e097e5e0493de033270a01b324caf31f464dc67'), new _limitordersdk.Address('0x90cbe4bdd538d6e9b379bff5fe72c3d67a521de5'), new _limitordersdk.Bps(10n), new _limitordersdk.Bps(6000n))), _index1.Whitelist.new(1738650226n, [
            {
                address: taker,
                allowFrom: 0n
            }
        ]));
        var calculatorNoSurplusFee = new _amountcalculator.AmountCalculator(auction, feeCalculator);
        var takingAmount = 100000n;
        var makingAmount = 1000n;
        var surplus = takingAmount / 3n;
        var estimatedTakingAmount = takingAmount - surplus;
        var calculator = new _amountcalculator.AmountCalculator(auction, feeCalculator, new _index1.SurplusParams(estimatedTakingAmount, _limitordersdk.Bps.fromPercent(50)));
        var userAmount = calculatorNoSurplusFee.getUserTakingAmount(taker, makingAmount, takingAmount, makingAmount, execTime);
        var userAmountWithChargedSurplus = calculator.getUserTakingAmount(taker, makingAmount, takingAmount, makingAmount, execTime);
        expect(userAmount).toBeGreaterThan(userAmountWithChargedSurplus);
        expect(userAmount - userAmountWithChargedSurplus).toEqual(surplus / 2n);
    });
}) // fee is 50%
;
