// @1inch/limit-order-sdk/dist/cjs/limit-order/extensions/extension-builder.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "ExtensionBuilder", {
    enumerable: true,
    get: function() {
        return ExtensionBuilder;
    }
});
var _byteutils = require("@1inch/byte-utils");
var _assert = /*#__PURE__*/ _interop_require_default(require("assert"));
var _extension = require("./extension.js");
var _constants = require("../../constants.js");
function _class_call_check(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}
function _defineProperties(target, props) {
    for(var i = 0; i < props.length; i++){
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
    }
}
function _create_class(Constructor, protoProps, staticProps) {
    if (protoProps) _defineProperties(Constructor.prototype, protoProps);
    if (staticProps) _defineProperties(Constructor, staticProps);
    return Constructor;
}
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
var ExtensionBuilder = /*#__PURE__*/ function() {
    "use strict";
    function ExtensionBuilder() {
        _class_call_check(this, ExtensionBuilder);
        _define_property(this, "makerAssetSuffix", _constants.ZX);
        _define_property(this, "takerAssetSuffix", _constants.ZX);
        _define_property(this, "makingAmountData", _constants.ZX);
        _define_property(this, "takingAmountData", _constants.ZX);
        _define_property(this, "predicate", _constants.ZX);
        _define_property(this, "makerPermit", _constants.ZX);
        _define_property(this, "preInteraction", _constants.ZX);
        _define_property(this, "postInteraction", _constants.ZX);
        _define_property(this, "customData", _constants.ZX);
    }
    _create_class(ExtensionBuilder, [
        {
            key: "withMakerAssetSuffix",
            value: function withMakerAssetSuffix(suffix) {
                (0, _assert.default)((0, _byteutils.isHexString)(suffix), 'MakerAssetSuffix must be valid hex string');
                this.makerAssetSuffix = suffix;
                return this;
            }
        },
        {
            key: "withTakerAssetSuffix",
            value: function withTakerAssetSuffix(suffix) {
                (0, _assert.default)((0, _byteutils.isHexString)(suffix), 'TakerAssetSuffix must be valid hex string');
                this.takerAssetSuffix = suffix;
                return this;
            }
        },
        {
            key: "withMakingAmountData",
            value: /**
     *
     * @param address Address of contract which will be called with `data` to calculate making amount
     * @param data
     */ function withMakingAmountData(address, data) {
                (0, _assert.default)((0, _byteutils.isHexString)(data), 'MakingAmountData must be valid hex string');
                this.makingAmountData = address.toString() + (0, _byteutils.trim0x)(data);
                return this;
            }
        },
        {
            key: "withTakingAmountData",
            value: /**
     *
     * @param address Address of contract which will be called with `data` to calculate taking amount
     * @param data
     */ function withTakingAmountData(address, data) {
                (0, _assert.default)((0, _byteutils.isHexString)(data), 'TakingAmountData must be valid hex string');
                this.takingAmountData = address.toString() + (0, _byteutils.trim0x)(data);
                return this;
            }
        },
        {
            key: "withPredicate",
            value: function withPredicate(predicate) {
                (0, _assert.default)((0, _byteutils.isHexString)(predicate), 'Predicate must be valid hex string');
                this.predicate = predicate;
                return this;
            }
        },
        {
            key: "withMakerPermit",
            value: function withMakerPermit(tokenFrom, permitData) {
                (0, _assert.default)((0, _byteutils.isHexString)(permitData), 'Permit data must be valid hex string');
                this.makerPermit = tokenFrom.toString() + (0, _byteutils.trim0x)(permitData);
                return this;
            }
        },
        {
            key: "withPreInteraction",
            value: function withPreInteraction(interaction) {
                this.preInteraction = interaction.encode();
                return this;
            }
        },
        {
            key: "withPostInteraction",
            value: function withPostInteraction(interaction) {
                this.postInteraction = interaction.encode();
                return this;
            }
        },
        {
            key: "withCustomData",
            value: function withCustomData(data) {
                (0, _assert.default)((0, _byteutils.isHexString)(data), 'Custom data must be valid hex string');
                this.customData = data;
                return this;
            }
        },
        {
            key: "build",
            value: function build() {
                return new _extension.Extension({
                    makerAssetSuffix: this.makerAssetSuffix,
                    takerAssetSuffix: this.takerAssetSuffix,
                    makingAmountData: this.makingAmountData,
                    takingAmountData: this.takingAmountData,
                    predicate: this.predicate,
                    makerPermit: this.makerPermit,
                    preInteraction: this.preInteraction,
                    postInteraction: this.postInteraction,
                    customData: this.customData
                });
            }
        }
    ]);
    return ExtensionBuilder;
}();
