// @1inch/fusion-sdk/dist/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    Address: function() {
        return _limitordersdk.Address;
    },
    AmountMode: function() {
        return _limitordersdk.AmountMode;
    },
    Extension: function() {
        return _limitordersdk.Extension;
    },
    ExtensionBuilder: function() {
        return _limitordersdk.ExtensionBuilder;
    },
    Interaction: function() {
        return _limitordersdk.Interaction;
    },
    LimitOrderContract: function() {
        return _limitordersdk.LimitOrderContract;
    },
    MakerTraits: function() {
        return _limitordersdk.MakerTraits;
    },
    OrderStatus: function() {
        return _index.OrderStatus;
    },
    Preset: function() {
        return _index.Preset;
    },
    PresetEnum: function() {
        return _index.PresetEnum;
    },
    Quote: function() {
        return _index.Quote;
    },
    QuoterCustomPresetRequest: function() {
        return _index.QuoterCustomPresetRequest;
    },
    QuoterRequest: function() {
        return _index.QuoterRequest;
    },
    RelayerRequest: function() {
        return _index.RelayerRequest;
    },
    TakerTraits: function() {
        return _limitordersdk.TakerTraits;
    },
    getLimitOrderContract: function() {
        return _limitordersdk.getLimitOrderContract;
    },
    getLimitOrderV4Domain: function() {
        return _limitordersdk.getLimitOrderV4Domain;
    },
    randBigInt: function() {
        return _limitordersdk.randBigInt;
    }
});
var _limitordersdk = require("@1inch/limit-order-sdk");
_export_star(require("./fusion-order/index.js"), exports);
_export_star(require("./amount-calculator//index.js"), exports);
_export_star(require("./connector/index.js"), exports);
_export_star(require("./sdk/index.js"), exports);
_export_star(require("./constants.js"), exports);
_export_star(require("./utils.js"), exports);
_export_star(require("./utils/amounts.js"), exports);
_export_star(require("./utils/time.js"), exports);
_export_star(require("./validations.js"), exports);
_export_star(require("./ws-api/index.js"), exports);
_export_star(require("./errors.js"), exports);
var _index = require("./api/index.js");
function _export_star(from, to) {
    Object.keys(from).forEach(function(k) {
        if (k !== "default" && !Object.prototype.hasOwnProperty.call(to, k)) {
            Object.defineProperty(to, k, {
                enumerable: true,
                get: function() {
                    return from[k];
                }
            });
        }
    });
    return from;
}
